### PR TITLE
feat(dgw): network scan HTTP API

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
Added basic api websocket endpoint for network scan

### CODE REVIEW

File: devolutions-gateway/Cargo.toml [hunk 1]
```diff
@@ -94,6 +94,7 @@ packet = { git = "https://github.com/fdubois1/rust-packet.git" }
 
 # For KDC proxy
 portpicker = "0.1"
+network-scanner = { version = "0.0.0", path = "../crates/network-scanner" }
```
Comment (CBenoit) [inline devolutions-gateway/Cargo.toml:97]: nitpick: This is unrelated to the KDC proxy, you can move this to the "in-house" section instead

File: devolutions-gateway/src/api/network_scan.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                        break;
+                    };
+
+                    if let Ok(Message::Close(_)) = msg {
+                        break;
+                    }
+                }
+            }
+        }
+        tracing::info!("Network scan finished");
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:58]: nitpick: tracing macros are all imported by default (see the #[macro_use] in lib.rs). Same at the other places.

```suggestion
        info!("Network scan finished");
```

File: devolutions-gateway/src/api/network_scan.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+#[derive(Debug, Deserialize)]
+pub struct NetworkScanQueryParams {
+    pub ping_interval: Option<u64>,     // in milliseconds,default 200
+    pub ping_timeout: Option<u64>,      // in milliseconds,default 500
+    pub broadcast_timeout: Option<u64>, // in milliseconds,default 1000
+    pub port_scan_timeout: Option<u64>, // in milliseconds,default 1000
+    pub netbios_timeout: Option<u64>,   // in milliseconds,default 1000
+    pub netbios_interval: Option<u64>,  // in milliseconds,default 200
+    pub max_wait: Option<u64>,          // max_wait for entire scan duration in milliseconds, suggested!
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:72]: nitpick: Use doc strings

```rust
    // Ping interval in milliseconds (default is 200)
    pub ping_interval: Option<u64>,
```

File: devolutions-gateway/src/api/network_scan.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+const FAMOUS_PORTS: [u16; 10] = [22, 23, 80, 443, 389, 636, 3389, 5900, 5985, 5986];
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:75]: naming: `COMMON_PORTS`

File: devolutions-gateway/src/api/network_scan.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct NetworkScanResponse {
+    pub ip: String,
+    pub hostname: Option<String>,
+    #[serde(rename = "type")]
+    pub type_: String,
+}
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:97]: suggestion: For `ip`, you can use the `IpAddr` type directly instead of `String`. `Serialize` is implemented on it.

naming: `ty` is a common abbreviation to avoid the alien trailing `_`

BUT, question: what about calling this field `protocol` or `application_protocol` (we use the word "application protocol" abbreviated to "ap" in our tokens)

suggestion: Maybe we can reuse the `Protocol` type found in the `token` module, instead of `String`.
This way, we also keep the same stringified representation at both places (e.g.: "ssh" instead of "SSH" and the "winrm-http-pwsh" / "winrm-https-pwsh" combination instead of WinRM)

File: devolutions-gateway/src/api/network_scan.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    response::Response,
+};
+
+use network_scanner::scanner::NetworkScannerParams;
+use serde::Serialize;
+
+use crate::{http::HttpError, DgwState};
+
+pub async fn handler(
+    _: State<DgwState>,
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:14]: question: Do we absolutely need to have this parameter? Seems unrequired, but maybe axum needs it for compilation.

File: devolutions-gateway/src/api/network_scan.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    ws: WebSocketUpgrade,
+    query_params: axum::extract::Query<NetworkScanQueryParams>,
+) -> Result<Response, HttpError> {
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:17]: issue: Seems like we are not checking for the NETSCAN token. See the `get_sessions` route for an example using an extractor.

File: devolutions-gateway/src/api/network_scan.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    let scanner_params: NetworkScannerParams = query_params.0.into();
+
+    let scanner = network_scanner::scanner::NetworkScanner::new(scanner_params).map_err(|e| {
+        tracing::error!("Failed to create network scanner: {:?}", e);
+        HttpError::internal().build(e)
+    })?;
+
+    let res = ws.on_upgrade(move |mut websocket| async move {
+        let stream = scanner.start().expect("Failed to start network scanner");
```
Comment (CBenoit) [inline devolutions-gateway/src/api/network_scan.rs:26]: issue: `expect` will panics on failure. If you can’t return an error, log an error instead.

**Review by CBenoit — COMMENTED** [review]
Thank you!

File: devolutions-gateway/Cargo.toml (hunk 1 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/Cargo.toml]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 2 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 4 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 5 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 7 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 8 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated

File: devolutions-gateway/src/api/network_scan.rs (hunk 6 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated, we don't need it, thanks!

File: devolutions-gateway/src/api/network_scan.rs (hunk 3 above)
Comment (irvingoujAtDevolution) [reply devolutions-gateway/src/api/network_scan.rs]: updated
